apiserver/logsink.go

```diff
@@ -154,6 +154,43 @@ type logSinkHandler struct {
 	fileLogger  io.Writer
 }
 
+// Since the logsink only recieves messages, it is possible for the other end
+// to disappear without the server noticing. To fix this, we use the
+// underlying websocket control messages ping/pong. Periodically the server
+// writes a ping, and the other end replies with a pong. Now the tricky bit is
+// that it appears in all the examples found on the interweb that it is
+// possible for the control message to be sent successfully to something that
+// isn't entirely alive, which is why relying on an error return from the
+// write call is insufficient to mark the connection as dead. Instead the
+// write and read deadlines inherent in the underlying Go networking libraries
+// are used to force errors on timeouts. However the underlying network
+// libraries use time.Now() to determine whether or not to send errors, so
+// using a testing clock here isn't going to work. So we rely on manual
+// testing, and what is defined as good practice by the library authors.
+//
+// Now, in theory, we should be using this ping/pong across all the websockets,
+// but that is a little outside the scope of this piece of work.
+
```

Comment:
I think a comment here about:

The server sends a Ping every 'pingPeriod' and when it receives a Pong it sets the socket Deadline to Now()+pongDelay. We chose pingPeriod so that pongDelay creates an overlap that has enough time for a successful response. eg. If we Ping at second 0, and get a Pong at second 1, we then set the deadline to 91s. The next ping at 60s then has 31s to get a response before the deadline would be reached.